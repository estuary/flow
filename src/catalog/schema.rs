use super::{sql_params, ContentType, Resource, Result, Scope, DB};
use crate::doc::{Schema as CompiledSchema, SchemaIndex};
use crate::specs::build as specs;
use estuary_json::schema::{build::build_schema, Application, Keyword};
use url::Url;

pub const INLINE_POINTER_KEY: &str = "ptr";

/// Schema represents a JSON-Schema document, and an optional fragment which
/// further locates a specific sub-schema thereof.
#[derive(Debug)]
pub struct Schema {
    pub resource: Resource,
    pub fragment: Option<String>,
}

impl Schema {
    /// Retrieve the primary URL of this Schema Resource, joined with the fragment sub-location.
    pub fn primary_url_with_fragment(&self, db: &DB) -> Result<Url> {
        let mut url = self.resource.primary_url(db)?;
        url.set_fragment(self.fragment.as_deref());
        Ok(url)
    }

    /// Register a JSON-Schema document relative to a current, external Resource
    /// (which must be the current Resource of the Scope). The JSON-Schema may
    /// be either a relative URL, or inline.
    /// * If a relative URL, the canonical URL is determined by joining with the
    ///   Scope's base URL in the usual way.
    /// * If an inline URL, a canonical URL is created by extending the Scope's
    ///   base URL with a query argument which captures the Scope's current
    ///   location. The inline document is then registered as a resource at that
    ///   (fictional) URL.
    pub fn register(scope: Scope, spec: &specs::Schema) -> Result<Schema> {
        // Map either inline vs relative URL cases to a canonical URL.
        let url = match spec {
            specs::Schema::Object(_) | specs::Schema::Bool(_) => {
                // Create the full URL by taking the URL of the parent resource (typically a
                // catalog yaml document) and adding a query parameter with a json pointer to
                // the location of the inline schema.
                let mut url = scope.resource().primary_url(scope.db)?;
                url.set_query(Some(&format!(
                    "{}={}",
                    INLINE_POINTER_KEY,
                    scope.location.url_escaped()
                )));

                Resource::register_content(
                    scope.db,
                    ContentType::Schema,
                    &url,
                    serde_json::to_vec(spec)?.as_slice(),
                )?;
                url
            }
            specs::Schema::Url(url) => scope.resource().primary_url(scope.db)?.join(url)?,
        };

        Self::register_url(scope, &url)
    }

    /// Register a JSON-Schema document at the URL, which must be the canonical
    /// URI of the schema. If already registered, this is a no-op and its existing
    /// handle is returned. Otherwise the document and all of its recursive references
    /// are added to the catalog.
    pub fn register_url(scope: Scope, canonical_url: &Url) -> Result<Schema> {
        // Schema URLs frequently have fragment components which locate a specific
        // sub-schema within the schema document. Decompose it to track separately,
        // then work with the entire schema document.
        let fragment = canonical_url.fragment().map(str::to_owned);
        let mut base_url = canonical_url.clone();
        base_url.set_fragment(None);

        let resource = Resource::register(scope.db, ContentType::Schema, &base_url)?;
        let scope = scope.push_resource(resource);

        let dom = resource.content(scope.db)?;
        let dom = serde_yaml::from_slice::<serde_json::Value>(&dom)?;
        let compiled: CompiledSchema = build_schema(resource.primary_url(scope.db)?, &dom)?;

        if !resource.is_processed(scope.db)? {
            resource.mark_as_processed(scope.db)?;

            // Walk the schema to identify sub-schemas having canonical URIs which differ
            // from the registered |url|. Each of these canonical URIs is registered as
            // an alternate URL of this schema resource. By doing this, when we encounter
            // a direct reference to a sub-schema's canonical URI elsewhere, we will
            // correctly resolve it back to this resource.
            Self::register_alternate_urls(scope, &compiled)?;

            // Walk the schema again, this time registering schemas which it references.
            // Since we've already registered alternate URLs, usages of those URLs
            // in references will correctly resolve back to this document.
            Self::register_references(scope, &compiled)?;
        }

        // Index the compiled schema, and confirm that the registered URL (with fragment)
        // resolves correctly.
        let mut index = SchemaIndex::new();
        index.add(&compiled)?;
        index.must_fetch(canonical_url)?;

        Ok(Schema { resource, fragment })
    }

    /// Walks compiled schema and registers an alternate URL for each encountered $id.
    fn register_alternate_urls(scope: Scope, compiled: &CompiledSchema) -> Result<()> {
        // Register Schemas having fragment-less canonical URIs.
        // Note the JSON-Schema spec requires that $id applications have no fragment.
        if compiled.curi.fragment().is_none() {
            scope
                .resource()
                .register_alternate_url(scope.db, &compiled.curi)?;
        }
        for kw in &compiled.kw {
            if let Keyword::Application(app, child) = kw {
                // Add Application keywords to the Scope's Location.
                let location = app.push_keyword(&scope.location);
                let scope = Scope {
                    location: app.push_keyword_target(&location),
                    ..scope
                };
                Self::register_alternate_urls(scope, child)?;
            }
        }
        Ok(())
    }

    /// Walks compiled schema and registers schemas which it references.
    fn register_references(scope: Scope, compiled: &CompiledSchema) -> Result<()> {
        for kw in &compiled.kw {
            if let Keyword::Application(app, child) = kw {
                // Add Application keywords to the Scope's Location.
                let location = app.push_keyword(&scope.location);
                let scope = Scope {
                    location: app.push_keyword_target(&location),
                    ..scope
                };
                // "Ref" applications indirect to a canonical schema URI which may
                // be in this document or another, and often include a fragment
                // component bearing a JSON-pointer into the document. We strip the
                // fragment here, since we're registering with whole-document granularity.
                if let Application::Ref(uri) = app {
                    let refed = Self::register_url(scope, uri)?;
                    Resource::register_import(scope, refed.resource)?;
                }
                // Recurse to sub-schemas.
                Self::register_references(scope, child)?;
            }
        }
        Ok(())
    }

    /*
    pub fn compile_required(db: &DB, collection: Collection) -> Result<Vec<CompiledSchema>> {
        use std::collections::HashSet;
        let mut resource_ids = HashSet::new();
        let mut current = collection.resource.id;

        // loop:
        //   - get schemas for this resource
        //   - get resources that this resource imports
        //       - for each recurse if not already fetched
    }

    fn compile_schemas_for_resource(
        db: &DB,
        resource_id: i64,
        visited: &mut HashSet<i64>,
    ) -> Result<Vec<CompiledSchema>> {
        let mut stmt = db.prepare(
            "SELECT url, content FROM resources NATURAL JOIN resource_urls
                WHERE resource_id = ? AND content_type = ? AND is_primary;",
        )?;
        let mut rows = stmt
            .query(sql_params![resource_id, ContentType::Schema])?
            .mapped(|row| (row.get::<String>(0)?, row.get::<Vec<u8>>(1)?));
        let mut schemas = Vec::new();

        for row in rows {
            let (url, content) = row?;

        }
    }
    */

    /// Fetch and compile all Schemas in the catalog.
    pub fn compile_all(db: &DB) -> Result<Vec<CompiledSchema>> {
        Self::compile_for(db, 1) // 1 is the root resource ID.
    }

    /// Fetch and compile all Schemas of the Resource, as well as Schemas it transitively imports.
    pub fn compile_for(db: &DB, resource_id: i64) -> Result<Vec<CompiledSchema>> {
        let mut stmt = db.prepare(
            "SELECT schema_uri, schema_content FROM resource_schemas WHERE resource_id = ?",
        )?;
        let mut rows = stmt.query(sql_params![resource_id])?;

        let mut schemas = Vec::new();
        while let Some(row) = rows.next()? {
            let (url, blob): (Url, Vec<u8>) = (row.get(0)?, row.get(1)?);
            let dom: serde_json::Value = serde_yaml::from_slice(&blob)?;
            let compiled: CompiledSchema = build_schema(url, &dom)?;
            schemas.push(compiled);
        }
        schemas.shrink_to_fit();
        Ok(schemas)
    }
}

#[cfg(test)]
mod test {
    use super::{
        super::{dump_table, dump_tables, init_db_schema, open},
        *,
    };
    use rusqlite::params as sql_params;
    use serde_json::{json, Value};

    #[test]
    fn test_register_with_alt_urls_and_self_references() -> Result<()> {
        let db = open(":memory:")?;
        init_db_schema(&db)?;

        let doc = json!({
            "$defs": {
                "wrapper": {
                    "$id": "test://example/root",
                    "$defs": {
                        "a": {
                            "$id": "test://example/other/a-doc",
                            "items": [
                                true,
                                {"$ref": "b-doc#/items/1"},
                            ],
                        },
                        "b": {
                            "$id": "test://example/other/b-doc",
                            "items": [
                                {"$ref": "a-doc#/items/0"},
                                true,
                            ],
                        },
                        "c": true,
                    },
                    "allOf": [
                        {"$ref": "other/a-doc#/items/1"},
                        {"$ref": "test://example/other/b-doc#/items/0"},
                        {"$ref": "#/$defs/c"},
                        {"$ref": "root#/$defs/c"},
                        {"$ref": "test://example/root#/$defs/c"},
                    ],
                }
            },
            "$ref": "test://example/root",
        });
        db.execute(
            "INSERT INTO resources (resource_id, content_type, content, is_processed) VALUES
                    (10, 'application/schema+yaml', CAST(? AS BLOB), FALSE);",
            &[doc],
        )?;
        db.execute(
            "INSERT INTO resource_urls (resource_id, url, is_primary) VALUES
                    (10, 'test://actual', TRUE);",
            sql_params![],
        )?;

        let url = Url::parse("test://actual")?;
        // Kick off registration using the registered resource path. Expect it works.
        let s = Scope::empty(&db).then(|scope| Schema::register_url(scope, &url))?;

        assert_eq!(s.resource.id, 10);
        assert!(s.resource.is_processed(&db)?);
        assert!(s.fragment.is_none());
        assert_eq!(s.primary_url_with_fragment(&db)?, url);

        assert_eq!(
            dump_table(&db, "resource_urls")?,
            json!([
                (10, "test://actual", true),
                (10, "test://example/root", Value::Null),
                (10, "test://example/other/a-doc", Value::Null),
                (10, "test://example/other/b-doc", Value::Null),
            ]),
        );

        Ok(())
    }

    #[test]
    fn test_register_with_external_references() -> Result<()> {
        let db = open(":memory:")?;
        init_db_schema(&db)?;

        let doc_a = json!({"$defs": {"a": {"$ref": "b#/$defs/c"}}});
        let doc_b = json!({"$defs": {"c": {"$ref": "c"}}});
        let doc_c = json!(true);
        let doc_d = json!(false);

        db.execute(
            "INSERT INTO resources
            (resource_id, content_type, content, is_processed) VALUES
            (10, 'application/schema+yaml', CAST(? AS BLOB), FALSE),
            (20, 'application/schema+yaml', CAST(? AS BLOB), FALSE),
            (30, 'application/schema+yaml', CAST(? AS BLOB), FALSE),
            (40, 'application/schema+yaml', CAST(? AS BLOB), FALSE);",
            sql_params![&doc_a, &doc_b, &doc_c, &doc_d],
        )?;
        db.execute(
            "INSERT INTO resource_urls (resource_id, url, is_primary) VALUES
                (10, 'file:///dev/null/a', TRUE),
                (20, 'file:///dev/null/b', TRUE),
                (30, 'file:///dev/null/c', TRUE),
                (40, 'file:///dev/null/d', TRUE);
            ",
            sql_params![],
        )?;
        let url = Url::parse("file:///dev/null/a#/$defs/a")?;
        let s = Schema::register_url(Scope::empty(&db), &url)?;
        assert_eq!(s.resource.id, 10);
        assert_eq!(s.fragment.as_ref().unwrap(), "/$defs/a");
        assert_eq!(s.primary_url_with_fragment(&db)?, url);

        assert_eq!(
            dump_tables(&db, &["resources", "resource_imports"])?,
            json!({
                "resources": [
                    [10, "application/schema+yaml", doc_a.to_string(), true],
                    [20, "application/schema+yaml", doc_b.to_string(), true],
                    [30, "application/schema+yaml", doc_c.to_string(), true],
                    [40, "application/schema+yaml", doc_d.to_string(), false], // Not reached.
                ],
                "resource_imports": [
                    [20, 30],
                    [10, 20],
                ],
            }),
        );

        Ok(())
    }

    #[test]
    fn inline_schema_is_registered() -> Result<()> {
        let db = open(":memory:")?;
        init_db_schema(&db)?;

        let catalog_yaml = r##"
            collections:
                - name: testcollection
                  schema:
                    type: object
                    additionalProperties: true
                  key: [/mykey]
                  fixtures: [nonexistant/fixtures.yaml]
        "##;

        db.execute(
            "INSERT INTO resources (resource_id, content_type, content, is_processed)
            VALUES
            (10, 'application/vnd.estuary.dev-catalog-spec+yaml', CAST(? AS BLOB), FALSE);",
            sql_params![catalog_yaml],
        )?;
        db.execute(
            "INSERT INTO resource_urls (resource_id, url, is_primary) VALUES
                (10, 'file:///dev/null/a/catalog.yaml', TRUE);",
            sql_params![],
        )?;

        let spec = serde_yaml::from_str::<specs::Catalog>(catalog_yaml)?;

        let result = Scope::empty(&db)
            .push_resource(Resource { id: 10 })
            .push_prop("path to")
            .push_item(32)
            .push_prop("schema")
            .then(|scope| Schema::register(scope, &spec.collections[0].schema))?;

        assert_eq!(
            Url::parse("file:///dev/null/a/catalog.yaml?ptr=/path%20to/32/schema")?,
            result.resource.primary_url(&db)?
        );
        assert!(result.fragment.is_none());

        let schema_content = result.resource.content(&db)?;
        let actual_schema: Value =
            serde_json::from_slice(schema_content.as_slice()).expect("failed to parse schema");

        let expected_schema = json!({
            "type": "object",
            "additionalProperties": true
        });
        assert_eq!(expected_schema, actual_schema);

        Ok(())
    }
}
