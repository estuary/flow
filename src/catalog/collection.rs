use super::{
    projections, sql_params, ContentType, Derivation, Resource, Result, Schema, Scope, DB,
};
use crate::specs::build as specs;

/// Collection represents a catalog Collection.
#[derive(PartialEq, Eq, Debug, Copy, Clone)]
pub struct Collection {
    pub id: i64,
    pub resource: Resource,
}

impl Collection {
    pub fn all(db: &DB) -> Result<Vec<Collection>> {
        let mut stmt = db.prepare("SELECT collection_id, resource_id FROM collections;")?;
        let rows = stmt.query(rusqlite::NO_PARAMS)?;
        rows.and_then(|row| {
            Ok(Collection {
                id: row.get(0)?,
                resource: Resource { id: row.get(1)? },
            })
        })
        .collect::<Result<Vec<_>>>()
    }

    /// Registers a Collection of the Source with the catalog.
    pub fn register(scope: Scope, spec: &specs::Collection) -> Result<Collection> {
        // Register and import the schema document.
        let schema = scope.push_prop("schema").then(|scope| {
            let schema = Schema::register(scope, &spec.schema)?;
            Resource::register_import(scope, schema.resource)?;
            Ok(schema)
        })?;

        scope
            .db
            .prepare_cached(
                "INSERT INTO collections (
                    collection_name,
                    schema_uri,
                    key_json,
                    resource_id
                ) VALUES (?, ?, ?, ?)",
            )?
            .execute(sql_params![
                spec.name,
                schema.primary_url_with_fragment(scope.db)?,
                serde_json::to_string(&spec.key)?,
                scope.resource().id,
            ])?;

        let collection = Collection {
            id: scope.db.last_insert_rowid(),
            resource: scope.resource(),
        };

        for (index, fixture) in spec.fixtures.iter().enumerate() {
            scope
                .push_prop("fixtures")
                .push_item(index)
                .then(|scope| collection.register_fixture(scope, fixture))?;
        }

        scope.push_prop("projections").then(|scope| {
            projections::register_projections(&scope, collection, &spec.projections)
        })?;

        if let Some(spec) = &spec.derivation {
            scope
                .push_prop("derivation")
                .then(|scope| Derivation::register(scope, collection, spec))?;
        }

        log::info!("added collection {}", spec.name);
        Ok(collection)
    }

    fn register_fixture(&self, scope: Scope, url: &str) -> Result<()> {
        let url = self.resource.join(scope.db, url)?;
        let fixtures = Resource::register(scope.db, ContentType::CatalogFixtures, &url)?;
        Resource::register_import(scope, fixtures)?;

        if !fixtures.is_processed(scope.db)? {
            scope.push_resource(fixtures).then(|scope| {
                // Just verify fixtures parse correctly.
                serde_yaml::from_slice::<Vec<specs::Fixture>>(&fixtures.content(scope.db)?)?;
                fixtures.mark_as_processed(scope.db)
            })?;
        }
        scope
            .db
            .prepare_cached("INSERT INTO fixtures (collection_id, resource_id) VALUES (?, ?)")?
            .execute(sql_params![self.id, fixtures.id])?;

        Ok(())
    }
}

#[cfg(test)]
mod test {
    use super::{
        super::{dump_tables, init_db_schema, open},
        *,
    };
    use rusqlite::params as sql_params;
    use serde_json::json;

    #[test]
    fn test_register() -> Result<()> {
        let db = open(":memory:")?;
        init_db_schema(&db)?;

        let schema = json!({
            "$anchor": "foobar",
            "type": "object",
            "properties": {
                "a": {
                    "type": "object",
                    "properties": {
                        "a": {
                            "type": "string"
                        }
                    }
                },
                "b": {
                    "type": "object",
                    "properties": {
                        "b": {
                            "type": "string"
                        }
                    }
                },
                "key": {
                    "type": "array",
                    "items": {
                        "type": "string"
                    },
                    "minItems": 2
                }
            }
        });
        let fixtures = json!([
            {
                "document": {"key": ["foo", "bar"], "other": "value"},
                "key": ["bar", "foo"],
                "projections": {"field-name": "value"}
            },
        ]);
        db.execute(
            "INSERT INTO resources (resource_id, content_type, content, is_processed) VALUES
                    (1, 'application/vnd.estuary.dev-catalog-spec+yaml', X'1234', FALSE),
                    (10, 'application/schema+yaml', CAST(? AS BLOB), FALSE),
                    (20, 'application/vnd.estuary.dev-catalog-fixtures+yaml', CAST(? AS BLOB), FALSE);",
            sql_params![schema, fixtures],
        )?;
        db.execute(
            "INSERT INTO resource_urls (resource_id, url, is_primary) VALUES
                    (1, 'test://example/spec', TRUE),
                    (10, 'test://example/schema.json', TRUE),
                    (20, 'test://example/fixtures.json', TRUE);",
            sql_params![],
        )?;

        let spec: specs::Collection = serde_json::from_value(json!({
            "name": "test/collection",
            "schema": "schema.json#foobar",
            "key": ["/key/1", "/key/0"],
            "fixtures": ["fixtures.json"],
            "projections": {
                "field_a": {"location": "/a/a", "partition": true},
                "field_b": {"location": "/b/b", "partition": false},
            }
        }))?;

        Scope::empty(&db)
            .push_resource(Resource { id: 1 })
            .then(|scope| Collection::register(scope, &spec))
            .expect("failed to register collection");

        // Expect that the schema and fixtures were processed.
        assert!(Resource { id: 10 }.is_processed(&db)?);
        assert!(Resource { id: 20 }.is_processed(&db)?);

        // Expect the collection records the absolute schema URI, with fragment component.
        let dump = dump_tables(
            &db,
            &[
                "resource_imports",
                "collections",
                "projections",
                "partitions",
                "fixtures",
            ],
        )?;

        assert_eq!(
            dump,
            json!({
                "resource_imports": [[1, 10], [1, 20]],
                "collections": [
                    [
                        1,
                        "test/collection",
                        "test://example/schema.json#foobar",
                        ["/key/1","/key/0"],
                        1,
                    ],
                ],
                "fixtures": [[1, 20]],
                "projections": [
                    [1, "field_a", "/a/a", true],
                    [1, "field_b", "/b/b", true],
                ],
                "partitions": [[1, "field_a"]],
            }),
        );

        Ok(())
    }
}
