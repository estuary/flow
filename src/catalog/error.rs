use super::{ContentType, ProjectionsError};
use crate::doc;
use estuary_json::schema;

#[derive(thiserror::Error, Debug)]
pub enum Error {
    #[error("at {loc}:\n{detail}")]
    At { loc: String, detail: Box<Error> },

    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),
    #[error("joining '{relative}' with base URL '{base}': {detail}")]
    URLJoinErr {
        base: url::Url,
        relative: String,
        detail: url::ParseError,
    },
    #[error("parsing URL: {0}")]
    URLParseErr(#[from] url::ParseError),
    #[error("HTTP error (reqwest): {0}")]
    ReqwestErr(#[from] reqwest::Error),
    #[error("failed to parse YAML: {0}")]
    YAMLErr(#[from] serde_yaml::Error),
    #[error("failed to parse JSON: {0}")]
    JSONErr(#[from] serde_json::Error),
    #[error("catalog database error: {0}")]
    SQLiteErr(#[from] rusqlite::Error),
    #[error("cannot fetch resource URI: {0}")]
    FetchNotSupported(url::Url),
    #[error(
        "{source_uri:?} references {import_uri:?} without directly or indirectly importing it"
    )]
    MissingImport {
        source_uri: String,
        import_uri: String,
    },
    #[error("{source_uri:?} imports {import_uri:?}, but {import_uri:?} already transitively imports {source_uri:?}")]
    CyclicImport {
        source_uri: String,
        import_uri: String,
    },
    #[error("resource has content-type {next}, but is already registered with type {prev}")]
    ContentTypeMismatch {
        next: ContentType,
        prev: ContentType,
    },
    #[error("failed to build schema: {0}")]
    SchemaBuildErr(#[from] schema::build::Error),
    #[error("schema index: {0}")]
    SchemaIndexErr(#[from] schema::index::Error),
    #[error("subprocess {process:?} failed with status {status}")]
    SubprocessFailed {
        process: std::path::PathBuf,
        status: std::process::ExitStatus,
    },
    #[error("{context}, location '{ptr}': {msg} @ schema '{schema_uri}'")]
    ExtractedFieldErr {
        schema_uri: String,
        ptr: String,
        context: String,
        msg: String,
    },
    #[error("schema validation error: {}", serde_json::to_string_pretty(.0).unwrap())]
    FailedValidation(doc::FailedValidation),
    #[error("Unable to generate default projections due to: {0}")]
    InvalidProjections(#[from] ProjectionsError),
}
