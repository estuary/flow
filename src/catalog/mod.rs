mod collection;
mod content_type;
mod db;
mod derivation;
mod error;
mod extraction;
mod lambda;
mod materialization;
mod nodejs;
mod projections;
mod regexp_sql_fn;
mod resource;
mod schema;
mod scope;
mod selector;
mod source;
pub mod specs;
mod test_case;
mod unicode_collation;

use std::path::Path;
use url::Url;

pub use collection::Collection;
pub use content_type::ContentType;
pub use derivation::Derivation;
pub use error::Error;
pub use lambda::Lambda;
pub use materialization::MaterializationTarget;
pub use resource::Resource;
pub use rusqlite::{params as sql_params, Connection as DB};
pub use schema::Schema;
pub use scope::Scope;
pub use selector::Selector;
pub use source::Source;
pub use test_case::TestCase;

pub type Result<T> = std::result::Result<T, Error>;

/// Create a new and empty catalog database, returning an open connection.
/// Any existing database at the given path is truncated.
pub fn create(path: &str) -> Result<DB> {
    if path != ":memory:" {
        // Create or truncate the database at |path|.
        std::fs::write(path, &[])?;
    }
    let c = open(path)?;
    c.execute_batch("BEGIN;")?;
    db::init(&c)?;
    c.execute_batch("COMMIT;")?;
    Ok(c)
}

/// Open an existing catalog database, which must already exist.
pub fn open<P: AsRef<std::path::Path>>(path: P) -> Result<DB> {
    let c = DB::open_with_flags(path, rusqlite::OpenFlags::SQLITE_OPEN_READ_WRITE)?;
    regexp_sql_fn::install(&c)?; // Install support for REGEXP operator.
    unicode_collation::install(&c)?;
    Ok(c)
}

pub use nodejs::build_package as build_nodejs_package;

const BUILD_COMPLETE_DESCRIPTION: &str = "completed catalog build";

/// Builds a catalog
pub fn build(db: &DB, spec_url: Url, nodejs_dir: &Path) -> Result<()> {
    db.execute_batch("BEGIN;")?;

    Source::register(Scope::empty(db), spec_url)?;
    extraction::verify_extracted_fields(db)?;
    build_nodejs_package(db, nodejs_dir)?;
    db.execute(
        "INSERT INTO build_info (description) VALUES (?);",
        rusqlite::params![BUILD_COMPLETE_DESCRIPTION],
    )?;

    db.execute_batch("COMMIT;")?;
    Ok(())
}

/// Returns true if the database `build_info` table indicates that a build was completed.
/// We don't do any sort of sophisticated up-to-date checks, since the ability to skip re-building
/// the database requires an explicit opt-in.
pub fn database_is_built(db: &DB) -> bool {
    let sql = "SELECT time FROM build_info WHERE description = ?";
    if let Ok(mut stmt) = db.prepare(sql) {
        let params = rusqlite::params![BUILD_COMPLETE_DESCRIPTION];
        if let Ok(timestamp) = stmt.query_row(params, |r| r.get::<usize, String>(0)) {
            log::debug!("database was built at: {}", timestamp);
            return true;
        }
    }
    false
}

// Not public; used for testing within sub-modules.
#[cfg(test)]
use db::test::{dump_table, dump_tables};

#[cfg(test)]
mod test {
    use std::env;
    use std::path::PathBuf;
    use std::process::Command;

    #[test]
    fn run_catalog_test() {
        let mut path = PathBuf::from(&env::var("CARGO_MANIFEST_DIR").unwrap());
        path.extend(&["src", "catalog", "test_catalog.sh"]);

        let status = Command::new(path.as_os_str())
            .spawn()
            .expect("failed to start test_catalog.sh")
            .wait()
            .expect("failed to wait for command");

        assert!(status.success());
    }
}
