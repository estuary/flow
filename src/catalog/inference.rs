use crate::catalog::{self, Schema, Scope, DB};
use crate::doc::inference::Shape;
use crate::doc::Pointer;
use estuary_json::schema::types;
use estuary_json::Location;
use url::Url;

/// Type information that's been inferred for a specific location within a schema.
#[derive(Debug)]
pub struct Inference<'a> {
    pub location_ptr: String,
    pub shape: &'a Shape,
    pub must_exist: bool,
}

impl<'a> Inference<'a> {
    pub fn locate_within(location_ptr: &str, shape: &'a Shape) -> Option<Inference<'a>> {
        if let Some((located, must_exist)) = shape.locate(&Pointer::from(location_ptr)) {
            Some(Inference {
                location_ptr: location_ptr.to_owned(),
                shape: located,
                must_exist,
            })
        } else {
            None
        }
    }
}

/// Traverses the shape and persists inferences for all of the "projectable" locations within it.
pub fn register_all(scope: Scope, schema_uri: &Url) -> catalog::Result<()> {
    let shape = Schema::shape_for(scope.db, scope.resource().id, schema_uri)?;

    for inference in get_inferences(&shape) {
        register_one(scope.db, schema_uri.as_str(), &inference)?;
    }
    Ok(())
}

/// Persists the given inference. This function is safe to call multiple times for the same
/// inference. This behavior was chosen because inferences may be generated from both user-provided
/// and automatically generated projections.
pub fn register_one(db: &DB, schema_uri: &str, inference: &Inference) -> catalog::Result<()> {
    let types_json = inference.shape.type_.to_json_array();
    let params = rusqlite::params![
        schema_uri,
        inference.location_ptr.as_str(),
        types_json,
        inference.must_exist,
        inference.shape.title,
        inference.shape.description,
        inference.shape.string.content_type.as_ref(),
        inference.shape.string.format.as_ref(),
        inference.shape.string.is_base64,
        inference.shape.string.max_length.map(usize_to_i64)
    ];
    // Using OR IGNORE here so that we don't return an error if the inference already exists.
    db.prepare_cached(
        "INSERT OR IGNORE INTO inferences (
                schema_uri,
                location_ptr,
                types_json,
                must_exist,
                title,
                description,
                string_content_type,
                string_format,
                string_content_encoding_is_base64,
                string_max_length
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?);",
    )?
    .execute(params)?;
    Ok(())
}

/// Traverses all of the "projectable" locations in the shape and returns a vec of `Inference`s.
pub fn get_inferences<'a, 'b>(shape: &'a Shape) -> Vec<Inference<'a>> {
    let mut fields = Vec::with_capacity(32);
    fill_inferences(Location::Root, shape, true, &mut fields);
    fields
}

fn fill_inferences<'a, 'b>(
    location: Location<'b>,
    shape: &'a Shape,
    must_exist: bool,
    inferences: &mut Vec<Inference<'a>>,
) {
    inferences.push(Inference {
        location_ptr: location.pointer_str().to_string(),
        shape,
        must_exist,
    });

    // Traverse sub-locations of this location when it takes an object
    // or array type. As a rule, children must exist only if their parent
    // does, the parent can *only* take the applicable type, and it has
    // validations which require that the child exist.

    for property in &shape.object.properties {
        fill_inferences(
            location.push_prop(&property.name),
            &property.shape,
            must_exist && shape.type_ == types::OBJECT && property.is_required,
            inferences,
        );
    }

    for (index, item_shape) in shape.array.tuple.iter().enumerate() {
        fill_inferences(
            location.push_item(index),
            item_shape,
            must_exist
                && shape.type_ == types::ARRAY
                && index < shape.array.min.unwrap_or_default(),
            inferences,
        );
    }

    // As an aide to documentation of repeated items, produce an inference
    // using '-' ("after last item" within json-pointer spec).
    if let Some(item_shape) = &shape.array.additional {
        fill_inferences(location.push_prop("-"), item_shape, false, inferences);
    };
}

fn usize_to_i64(unsigned: usize) -> i64 {
    unsigned.min(usize::MAX - 1) as i64
}

#[cfg(test)]
mod test {
    use super::*;
    use estuary_json::schema::{build::build_schema, index::Index};
    use itertools::Itertools;

    #[test]
    fn inferences_are_registered_by_schema_uri() {
        let db = catalog::create(":memory:").unwrap();

        let schema = r##"
            type: object
            properties:
              a:
                type: object
                properties:
                  a: {type: string}
                  b: {type: boolean}
                  c:
                    type: integer
                    title: the C title
                    description: the C description
                  d: {type: number}
                required: [a, b]
            required: [a]
            "##;
        db.execute(
            "INSERT INTO resources (resource_id, content_type, content, is_processed) VALUES
                    (1, 'application/vnd.estuary.dev-catalog-spec+yaml', X'1234', FALSE),
                    (22, 'application/schema+yaml', CAST(? AS BLOB), FALSE);",
            rusqlite::params![schema],
        )
        .unwrap();
        db.execute_batch(
            r##"
            INSERT INTO resource_urls (resource_id, url, is_primary) VALUES
                    (1, 'test://test/flow.yaml', TRUE),
                    (22, 'test://schema.yaml', TRUE);
            INSERT INTO resource_imports (resource_id, import_id) VALUES (1, 22);
                         "##,
        )
        .unwrap();

        let uri = Url::parse("test://schema.yaml").unwrap();
        let scope = Scope::for_test(&db, 1);
        register_all(scope, &uri).expect("failed to register inferences");

        let actual = catalog::dump_table(&db, "inferences").unwrap();
        insta::assert_json_snapshot!(actual);
    }

    #[test]
    fn inferences_are_returned_for_shape() {
        let schema = r##"
            type: object
            properties:
              a:
                type: object
                properties:
                  a: {type: string}
                  b: {type: boolean}
                  c: {type: integer}
                  d: {type: number}
                required: [a, b]
              b:
                oneOf:
                  - type: object
                    properties:
                      a: {type: string}
                      b: {type: [boolean, "null"]}
                      c: {type: integer}
                      d:
                        type: array
                        allOf:
                            - items: [{type: integer}]
                              additionalItems: {type: [number, boolean]}
                            - items: {type: [number, string]}
                    required: [a, d]
                  - type: string

              c:
                type: array
                oneOf:
                    - items:
                        - type: string
                        - type: integer
                    - items: {type: "null"}
                minItems: 1
              d: {type: integer}
            required: [a, b, c]
            "##;
        let dom: serde_json::Value =
            serde_yaml::from_str(schema).expect("failed to compile schema");
        let schema_uri = url::Url::parse("test://test.test/test.json").unwrap();
        let compiled_schema =
            build_schema(schema_uri.clone(), &dom).expect("failed to compile schema");
        let mut index = Index::new();
        index.add(&compiled_schema).unwrap();
        let shape = Shape::infer(&compiled_schema, &index);

        let result = get_inferences(&shape);
        // Format the list of inferences into a string representation that will be easy
        // to view and compare as an insta snapshot.
        let repr = result
            .into_iter()
            .map(|inference| {
                let types = inference.shape.type_.to_json_array();
                format!(
                    "location: {:?}, types: {}, must_exist: {}",
                    inference.location_ptr, types, inference.must_exist
                )
            })
            .join("\n");
        insta::assert_snapshot!(repr);
    }
}
