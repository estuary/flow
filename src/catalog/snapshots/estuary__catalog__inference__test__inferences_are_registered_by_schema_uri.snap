---
source: src/catalog/inference.rs
expression: actual
---
[
  [
    "test://schema.yaml",
    "",
    [
      "object"
    ],
    true,
    null,
    null,
    null,
    null,
    null,
    null
  ],
  [
    "test://schema.yaml",
    "/a",
    [
      "object"
    ],
    true,
    null,
    null,
    null,
    null,
    null,
    null
  ],
  [
    "test://schema.yaml",
    "/a/a",
    [
      "string"
    ],
    true,
    null,
    null,
    null,
    null,
    null,
    null
  ],
  [
    "test://schema.yaml",
    "/a/b",
    [
      "boolean"
    ],
    true,
    null,
    null,
    null,
    null,
    null,
    null
  ],
  [
    "test://schema.yaml",
    "/a/c",
    [
      "integer"
    ],
    false,
    "the C title",
    "the C description",
    null,
    null,
    null,
    null
  ],
  [
    "test://schema.yaml",
    "/a/d",
    [
      "number"
    ],
    false,
    null,
    null,
    null,
    null,
    null,
    null
  ]
]
